test/helpers/envtest.go

```diff
@@ -21,95 +21,106 @@ import (
 	root := path.Join(path.Dir(filename), "..", "..")
 
-	return &TestEnvironment{
-		Environment: &envtest.Environment{
-			ErrorIfCRDPathMissing: true,
-			CRDDirectoryPaths: []string{
-				filepath.Join(root, "config", "crd", "bases"),
-				filepath.Join(root, "controlplane", "kubeadm", "config", "crd", "bases"),
-				filepath.Join(root, "bootstrap", "kubeadm", "config", "crd", "bases"),
-			},
-			CRDs: []runtime.Object{
-				external.TestGenericBootstrapCRD.DeepCopy(),
-				external.TestGenericBootstrapTemplateCRD.DeepCopy(),
-				external.TestGenericInfrastructureCRD.DeepCopy(),
-				external.TestGenericInfrastructureTemplateCRD.DeepCopy(),
-			},
+	// Create the test environment.
+	env = &envtest.Environment{
+		ErrorIfCRDPathMissing: true,
+		CRDDirectoryPaths: []string{
+			filepath.Join(root, "config", "crd", "bases"),
+			filepath.Join(root, "controlplane", "kubeadm", "config", "crd", "bases"),
+			filepath.Join(root, "bootstrap", "kubeadm", "config", "crd", "bases"),
+		},
+		CRDs: []runtime.Object{
+			external.TestGenericBootstrapCRD.DeepCopy(),
+			external.TestGenericBootstrapTemplateCRD.DeepCopy(),
+			external.TestGenericInfrastructureCRD.DeepCopy(),
+			external.TestGenericInfrastructureTemplateCRD.DeepCopy(),
 		},
-		scheme:  scheme,
-		doneMgr: make(chan struct{}),
 	}
 }
 
-func NewTestEnvironment() (*TestEnvironment, error) {
-	env := newTestEnvironment()
+// TestEnvironment encapsulates a Kubernetes local test environment.
+type TestEnvironment struct {
+	manager.Manager
+	client.Client
+	Config *rest.Config
 
-	cfg, err := env.Environment.Start()
-	if err != nil {
-		return env, err
+	doneMgr chan struct{}
+}
+
+// NewTestEnvironment creates a new environment spinning up a local api-server.
+//
+// This function should be called only once for each package you're running tests within,
+// usually the environment is initialized in a suite_test.go file within a `BeforeSuite` ginkgo block.
+func NewTestEnvironment() *TestEnvironment {
+	if _, err := env.Start(); err != nil {
+		panic(err)
 	}
 
-	mgr, err := manager.New(cfg, manager.Options{
-		Scheme:             env.scheme,
+	mgr, err := manager.New(env.Config, manager.Options{
+		Scheme:             scheme.Scheme,
 		MetricsBindAddress: "0",
 		NewClient:          util.ManagerDelegatingClientFunc,
 	})
 	if err != nil {
-		if stopErr := env.Stop(); err != nil {
-			return nil, kerrors.NewAggregate([]error{err, stopErr})
-		}
-		return env, err
+		klog.Fatalf("Failed to start testenv manager: %v", err)
 	}
 
-	env.Config = cfg
-	env.Manager = mgr
-	env.Client = mgr.GetClient()
-
-	return env, nil
+	return &TestEnvironment{
+		Manager: mgr,
+		Client:  mgr.GetClient(),
+		Config:  mgr.GetConfig(),
+		doneMgr: make(chan struct{}),
+	}
 }
 
 func (t *TestEnvironment) StartManager() error {
 	return t.Manager.Start(t.doneMgr)
 }
 
 func (t *TestEnvironment) Stop() error {
-	close(t.doneMgr)
-	return t.Environment.Stop()
+	t.doneMgr <- struct{}{}
```

Comment:
Does this actually stop the Manager? It looks like based on the docs that the manager is expected to stay running until the channel is closed: https://godoc.org/sigs.k8s.io/controller-runtime/pkg/manager#Runnable